### ISSUE TITLE
Support cold DOM event streams in templates

### ISSUE THREAD

**Issue by mhevery** [angular/angular#13248]

**Comment by mhevery** [comment]
# Requirements
* [R1] Observables must be cold (ie if there is no subscription then there should be no listeners.)
USE CASE: drag and drop. No need to listen on mousemove if drag has not been initiated.
* [R2] It must be possible to execute an expression which computes the action which will be next-ed into the observable.
  - USE CASE: ngFor shows a list of todos with a remove button. Clicking the remove should notify the stream not with the DOM event, but with the item to be removed.
* [R3] It must be possible to set up observable recipes to run inside/outside of angular zone.
  - USE CASE: 
    - drag and drop. Don't run CD on mousemove. 
    - Throttle; debounce; etc..: Only run CD on subscription which pass throttle, debounce; etc.
* [R4] It would be strongly preferable if there was something in the template which would indicate that an observable is listening at a given location. 
  - USE CASE: it should be easy for the developer/tooling to see where in the source code the event will be processed.

**Comment by frederikschubert** [comment]
Why can't we reuse the `|` combinator? As in `(event)="actionExp | stream"`.
This would then read:
In case of event X, pipe this Object into that stream. The concept of a pipe is pretty similar to an observable and the syntax fits nicely into the existing template syntax IMO.

**Comment by laco0416** [comment]
For the parser (and humans), I think a specific prefix symbol is needed.
AFAIK, RxJS users prefer `$` character to indicate that is `$(s)tream`.  So my idea is the below

- `<button ($click+remove)="{item: user, index: i}">X</button>`

**Comment by FabioAparicio** [comment]
The most useful thing (for the developer) about expressing events as streams is the ability to control the flow of events using operators like debounce etc.
To accomplish this all we need to do is create an emitter (Subject) to capture those events.

Something like
```
@Component({
  template: `
    <ul>
      <li *ngFor="let user of users; let i = index">
        {{user.name}}
        <button (click)="remove$.next({item: user, index: i})">X</button>
        <button (click=>remove)="remove$.next({item: user, index: i})">X</button>
      <li>
    <ul>
    `
})
class MyComponent implements OnInit {
  users: User[] = ...;
  remove$: Subject<{item: User, index: number}>;
  actions$: Observable<Action>;

  ngOnInit() {
    // An example of building up a recipe.
    // Notice that we are making sure that Rx operators have no CD.
    this.actions$ = noChangeDetection(() => this.remove$.map(() => ...));
  }
}
```

That said, there is indeed an opportunity to decrease the number of listeners by having lazyness, therefore reducing memory pressure.

As we know the same behaviour can be achieved with manual wiring up of event listeners using `Observable.fromEvent()` so the syntax has to be more convenient than manual wiring, otherwise it's not worth it.


The most familiar way would be to have something like:
`<button (click$remove)="{item: user, index: i}">X</button>`:

A hot stream:
`<button (click$remove.hot)="{item: user, index: i}">X</button>`
and a cold one:
`<button (click$remove.cold)="{item: user, index: i}">X</button>`
`<button (click$remove)="{item: user, index: i}">X</button>` (implicitly cold)

I think that is what conforms best with the current syntax, however I find that this would be much more terse:
`<button (click$)="remove({item: user, index: i})">X</button>`
But, it would probably require special casing the parser, doesn't seem viable.

If the microsyntax without a macrosyntax issue turns out not to be a problem then maybe this:
`<button ($click)="into remove using {item: user, index: i}">X</button>`
Which is verbose but explicit, a more terse option:
`<button ($click)="{item: user, index: i} into remove">X</button>`

**Comment by vicb** [comment]
Some thoughts on the proposed syntax:
- :-1: on `@Inject()` - "why can we Inject anything in the ctor but not in a property" ? Different mechanisms need different name 
- :-1: on the stream name on the LHS - non stream syntax is `(event)="action()"`, RHS seems more consistent in this sense.

What about `(*event)="obs.next(value)"` - The `*` is preferred **(not necessary)** as it hints that it is cold (not always listening)

_(btw, I don't get why the `*` hints it's cold ? could you expand ?)_

**Comment by vicb** [comment]
@mhevery could you please turn this issue into a proper spec (google docs) ? and keep this issue updated with the change log of the doc.

The doc should contain:
- Issue statement,
- Reqs,
- Proposal,
- Alternate proposals with pros and cons,
- Syntax,
- Use cases (implementation with the proposal),
- Change log.

I think it will be easier to discuss that way

**Comment by mhevery** [comment]
Just leaving for the record here all of the syntax forms which we have considered:

```
<input #a (keyup)="$event >> changes">
{{a.value}}

<ol *ngFor="let item of items">
<template ngFor let-item [ngForOf]="items">
  <button (*click|action)="{action: remove, item: item}" let-foo="..">
  <button stream-click-into-action="{action: remove, item: item}">
  
  <my-component (*open|action)="{action: 'open', item: $event}">

  <button (click)="{action: remove, item: item} >> action">
  <button (click)="{action: remove, item: item}">
  <button (*click)="{action: remove, item: item} >> action">
  <button (*click)="action << {action: remove, item: item}">
  <button (*click)="{action: remove, item: item} >> action().foo">
  
  
  <button (*click)="stream action << {action: remove, item: item}">
  <button (*click|action)="{action: remove, item: item}">
  <button on-click-into-action="{action: remove, item: item}">
  <button stream-click="{action: remove, item: item} >> action">
  <button stream-click="stream action action << {action: remove, item: item} >> action">
  <button (!click|action)="{action: remove, item: item}">
  <button {click|action}="{action: remove, item: item}">


class MyComp {
@Stream???()
action: Observable<{action:string, item: any}>;

}

foo() {}







<input #a (change)="exp()" (*keyup)="foo(); a.value >> change" >
<div (mousemove)="dragging ? a >> move : b"
     (*mousemove|onMove)="$event"
{{a.value}}



  
  
@Inject()
Stream<Item> removeStream
```

**Comment by mhevery** [comment]
@vicb and I had a brainstorming session and came up with this syntax. /cc @IgorMinar @blesh .  

```
<button (*click)="streamExpression >> dstStream">...</button>
```
Where:
-  `dstStream` is a property of the controller object which receives the stream.
- `streamExression` is something which adds operators to a stream such as `$stream.map(...).debounce()`

A common usage may be this:
```
@Component({
  template: 
    `<button (*click)="$stream.map(()=> {action:'add', text: 'test'}) >> action">test</button>`
})
class MyComponent {
  @Stream()
  action: Observable<{action: string}>;
}
```

The above syntax is a bit verbose, so we will have a special map syntax as it is the most used primitive with Observables.

```
<button (*click|action)="{action:'add', text: 'test'}">test</button>
```

It is important to understand when the RHS of the expression executes in the component lifecycle.

Given:
```
<ul>
  <li *ngFor="let item of items">
    <button (*click)="$stream.debounce(10).map(() => {action: 'REMOVE', item: item) >> action">remove</button>
  </li>
</ul>
<button (*click|action)="{action: 'ADD', text: 'test'}">add</button>
```

It is important to understand that the expression associated with `(*click)` will execute exactly once at the beginning of the template creation as the `*ngFor` is unrolled. This means that the only way to read the value `item` is through a closure, because the item can change at later point in time, and we want the current value of item not initial value of item

NOTE: (Closure syntax would have to be added to the parser)


## Pipes

The advantage of this syntax is that one cane use pipe and custom function to build up stream recipes such as:

```
<button (*click)="myFn($stream) | debounce:100 >> action">test</button>
```

**Comment by mhevery** [comment]
Unifying cold/hot streams.

```
@Component({
  template: `
    <!-- What can be done today (But event is hot registered)-->
    <button (click)="action.next($event)">
    <!-- Would be equivalent to simply piping `$stream` into `action`.-->
    <!-- But in this example the DOM event is registered only if a stream has listener -->
    <button (*click|action)="$stream">
    <!-- Which can be expressed in the short form as this -->
    <button (*click|action)>


    <!--  More important use case is nexting a value other then an event. -->
    <button (click)="action.next({action: 'ADD', value: item})">
    <!--  Which is the same as using the map operator on the observable. -->
    <button (*click|action)="$stream.map(() => {action: 'ADD', value: item})">
    <!--  Which can be simplified for the common form using AST transforms as such. -->
    <button (*click|action)="map: {action: 'ADD', value: item}">
  `
})
class MyComponent {
  // We instantiate the EventEmitter synchronously
  action: Observable<any> = new EventEmitter();

  // which then allows building up recipes like so
  removes = () => this.action.map(...));
}
```

Under the hood the when we compile a template:
```
<button (*click|action)="$stream.map(() => {action: 'ADD', value: item})">
```
It generates something like so:
```
////// GENERATED
  
ngOnViewInsert() {
  this.click1$ = $stream.map(() =>  {action: 'ADD', value: item});
  this.action.addStream(this.click1$);
}
  
ngOnViewRemove() {
  this.action.removeStream(this.click1$);  
}
```

**Comment by blesh** [comment]
My only concern with putting too much of your observable configuration in the view/template would be that it's going to be much harder to test. What if that set up was done in the component?

```ts
@Component({
  template: `
    <button (*click|action)="setupClicks">Click Me</button>
  `
})
class MyComponent {
  action: Observable<ClickAction> = null;

  i = 0;

  setupClicks(click$: Observable<MouseEvent>) {
    return click$.map(e => ({ type: 'CLICK_ACTION', value: this.i++ });
  }
}
```

This way you can unit test that setup function.

Seems like just as much boilerplate as anything though.

**Comment by vicb** [comment]
@blesh that would be supported with the proposed syntax:

```ts
@Component({
  template: `
    <button (*click|action)="setupClicks($stream)">Click Me</button>
  `
})
```

You want to put the logic in your template when you access context specific info (think loop item, index in a `NgFor`) which are not available on your component instance.

**Comment by wardbell** [comment]
I am intrigued but, I think like @blesh, I find so much syntax in the html to be smelly. Testing is one concern but it goes way beyond that.

Maybe when I read an earlier part of this thread I'll understand what use cases make coding the template so much better than coding in the component.  We will certainly have to articulate that to a skeptical world.

**Comment by mhevery** [comment]
Self Note:

Turns out that we can't use `foo:` as prefix for Ast transform as it would collide with macro-syntax. For example if we had: 

`[tooltip]=" 'text', height: 10 "`

 which would become 

`[tooltip]="foo: 'text', height: foo: 10"`

Which would create an ambiguity, as it would not be clear if we are binding to `tooltip` with `foo` AST transform or binding to `tooltipFoo` with no AST transform. So we need a different syntax. Perhaps something like `foo::` which would then become 

`[tooltip]="foo::'text' , height: foo::10  "`

or in the case of stream:

`<button (*click|action)="map::{action: 'ADD', value: item}">`

**Comment by fxck** [comment]
> AFAIK, RxJS users prefer $ character to indicate that is $(s)tream. So my idea is the below

Yes, they do, but appended, not prepended, `$whatever` usually meant jquery object, while `whatever$` is usually a stream.

**Comment by kylecordes** [comment]
I like to echo several others here with a recommendation to keep the template syntax minimal, just enough hook to wire up to the plain old TypeScript code in the class, keep the complexity over there.

**Comment by blesh** [comment]
It seems to me like something more minimalistic might work, something like:

```
@Component({
  template: `
    <input $q (keypress)="({ action: 'FOO', value: q.value }) -> qChange$"/>
    <div *ngFor="let result of searche$ | async">{{result.text}}</div>
  `
})
class MyComponent {
  qChange$: Observable<string>; 

  get searche$(): Observable<SearchResult[]> {
    this.qChange$.debounceTime(500)
      .switchMap(query =>
        http.get('/search/' + query)
          .map(res => res.json())
      )
  }
}
```

The `->` operator would be of the lowest possible precedence so `<stuff here>` in `<stuff here> -> observable$` would be processed *first*. It could even be something like `qChange$(<stuff here>)`, which might look more idiomatic to JS developers.

I really don't think it's a great idea to put too much in the way of composed operators in the template. For one thing, testing anything async that happens in there will become more difficult. A debounce, for example.

However, I can totally see why it would be beneficial to have some cleaner way to push into a cold observable.

**Comment by mhevery** [comment]
@blesh sorry but `(keypress)` would mean that the registration is hot. We need different semantics for cold registration. 

BTW, this works today.

`<input #q (keypress)="action.next({action: 'FOO', value: q.value })"`

**Comment by mhevery** [comment]
> Why is it not?
> 
> (click$|action)

It would be its the stream which has `$` not event, since one could have a component with such an event. So it would be `(click|action$)` if you name your stream that way. 

the reason for `*` is that we need some sort of prefix so that IDE tooling can work better. You want to notify the IDE before you autocomplete not after. It needs to be a char which could not be in variable name. So `$` or `_` is out.

**Comment by blesh** [comment]
> @blesh sorry but (keypress) would mean that the registration is hot.

@mhevery why is `keypress` hot, but `click` is not? Presumably they're both event streams registered in the same manner against a DOM element that doesn't exist until it's mounted to the view. Or do you mean I should put an `*` up front or something to signify that it's an event that isn't processed until mount-time. like `(*keypress)=""`?

**Comment by vicb** [comment]
@blesh because `(keypress)="..."` is an existing syntax in ng2 and uses `addEventListener()` while `(*click|dst)="..."` would be a new Angular construct and would use `Observable.fromEvent()`.

I think this GH issue mixes two concerns:
- how to express stream transformations, ie `(*src|dst)="$stream.trans1().trans2()"`,
- the fact that `src` streams built from DOM events should be cold - which I think nobody here disputes.

**Comment by kylecordes** [comment]
How about the following alternative, even less new syntax than described by @blesh?

In this approach, we would use the existing # to mark an element, and the existing () syntax to capture an event, but leave out the value for the attribute.

Although there seems to be a lot of enthusiasm here for lots of new syntax to use inside a template, overall I believe we are mostly better served by not introducing more new syntax or more kinds of expressions into the template language. We already have a very serviceable language, TypeScript, with a very serviceable way to annotated with extra bits that don't fit in the language itself.


```
@Component({
  template: `
    <input #foo (change) >
    <div *ngFor="let result of searche$ | async">{{result.text}}</div>
  `
})
class MyComponent {
  @Stream('foo', 'change')
  fooChange$: Observable<string>; 

  get searche$(): Observable<SearchResult[]> {
    this.fooChange$.debounceTime(500)
      .map(e => e.value)
      .switchMap(query =>
        http.get('/search/' + query)
          .map(res => res.json())
      )
  }
}
```

**Comment by blesh** [comment]
@vicb Then I'd stick with `(*src | propName)` or any other forward-compatible syntax and leave the transformation out of the proposal for now. /@mhevery

**Comment by gdi2290** [comment]
Can we use `^` rather than `*` since teaching `*` is already confusing enough. 

Rather than `@Stream` can we call it `@ObservableEvent` or something along those lines. 

Can we also change `ngOnCreate` a bit to be invoked before setting the event values so

```typescript
  @Input() value = DEFAULT_VALUE;
  ngOnCreate(observableInputs) {
    console.log(this.value) // DEFAULT_VALUE
  }
```
I need this for Universal to overwrite the default values during server/client hydration

```typescript
  constructor(public store: Store) {
  }
  @Input() value = DEFAULT_VALUE;
  ngOnCreate(observableInputs) {
    console.log(this.value) // DEFAULT_VALUE
    this.value = this.store.get('HomeComponent.value', this.value);
    // angular bootstraps with the batch data made on the server which allows the client to load faster since only one http call (fetching index.html) was made
  }
```
but also for the prototype made in this comment https://github.com/angular/angular/issues/4062#issuecomment-231671716 where we need the initial value to correctly implement a "mini-redux" built into angular
```typescript
  @Output() counterChange = Observable.merge(
      this.increments,
      this.decrements.mapTo(-1)
    )
    .startWith(this.counter)
    .scan((total, value) => total + value, 0); // notice the initial value is hard coded to 0 rather than using observableInputs
```

we can also expand the arguments for `ngOnCreate` which would lead to more CycleJS design (which is great)

```typescript
  ngOnCreate(componentLifeCycleRef: ComponentLifeCycleRef) {
    componentLifeCycleRef.inputs.value$ // object of input observable for the above examples
    componentLifeCycleRef.onDestory$ // we can now use takeUntil so we can avoid using ngOnDestory to unsub
    componentLifeCycleRef.onChanges$ // we can hook into cd
    componentLifeCycleRef.onInit$ // we won't need this as we get the values from inputs.value etc
  }
```
now that is extremely powerful

**Comment by mattpodwysocki** [comment]
@gdi2290 why is `0` hard coded as a seed as it won't be used since you're already using `startWith`?

**Comment by gdi2290** [comment]
@mattpodwysocki `0` is hard coded because we need an Observable of the Inputs (I hardcoded it because the example would look a bit more complex, see below). So the example was showing what we could do while the correct solution would look a bit weird without better framework support for example using `angular2-observe-decorators`
```typescript
  @ObserveInput() initialValue$ = new EventEmitter();
  @Output() counterChange = initialValue$.mergeMap(initialValue => {
    return Observable.merge(
      this.increments,
      this.decrements.mapTo(-1)
    )
    .startWith(this.counter)
    .scan((total, value) => total + value, initialValue);
  })

```
the observe-decorators create a proxy (Subjects) that will emit the values when they are available later in the lifecycle of the component. This will allow us to wire up the data flow when we create the component but it's a bit awkward. Having `ngOnCreate` that provided an ObservableComponentRef with Observables would help with this so we can do stuff like `.takeUntil` or grab the initial values

**Comment by gdi2290** [comment]
Another great thing about having `ngOnCreate` is that the separation of concerns for getting data in a component is more intuitive. Right now users are assuming they should grab data in the constructor which makes it hard to test and feels awkward.
```typescript
  constructor(public http: Http) {
    this.person = this.http.get('data.json')
     .map(res => res.json())
     .toPromise()
     .then(data => this.data = data);
  }
```
If we simply suggest users to use `ngOnCreate` for data then any future refactors would make a lot more sense. For example refactoring the data fetching to the router then by the time you inject the router in the component you will have the initial data wouldn't be too hard refactoring in ngOnCreate. Adding any sort of **Relay** like state management would also need to use `ngOnCreate` to set the initial values and to trigger change detection or not via helper function `noChangeDetection/changeDetection`

**Comment by blesh** [comment]
AAAHHH... `toPromise`!!! My eyes!

I think the `componentLifeCycleRef` is a great idea. Although the API seems a little broad (perhaps it's necessary). Is this something that already exists? (Sorry, I have my head in Falcor and RxJS these days). If not, I'd suggest implementing it in some more minimal way. Perhaps just a few of those properties mentioned, the ones that are highest value. If lessons are going to be taken from CycleJS, then I'd definitely talk to @staltz about his experiences implementing this. He might have some successes and regrets to share that would benefit you.

**Comment by fxck** [comment]
Any development on this? Looks like it's blocked on rxjs support for zonejs or something?

**Comment by xsilen** [comment]
more clear right now

**Comment by mitCode** [comment]
Is this going to be available in Angular 4 or is it put off to a future time? Couldn't find any info about it.

**Comment by DzmitryShylovich** [comment]
> put off to a future time

**Comment by fxck** [comment]
Any news `on` this @mhevery ? I heard some rumors, but it would be nice to have an officlal update

**Comment by wesamco** [comment]
Bump!
Go ahead thumbsdown vote/react 😆 

Jk aside, I think at least sometimes Engineers should prioritize features & projects that help with Google's revenue, like this feature & the [AngularFire2](https://github.com/angular/angularfire2) project. synergy between Angular and Firebase projects/teams should be encouraged.

apologies for bothering, if the [Core Roadmap](https://github.com/angular/angular/projects/6) with issues like [this one](https://github.com/angular/angular/issues/13725) are blocking the way, and that's why it's `put off to a future time`.

**Comment by fxck** [comment]
Hi, @robwormald, @mhevery could you please give an update on this?

**Comment by shyamal890** [comment]
This doesn't yet support `debounce` `ngModelOption` on inputs. I would like to know what is the logic behind not implementing the same with `blur`?

Example usecase for `debounce`:
- A text input like in Google docs which should save the typed text after 3 secs of no keypress after typing.

**Comment by LinBoLen** [comment]
I don't think it will be implemented currently. programming on template relate to rxjs?

**Comment by fxck** [comment]
https://github.com/whatwg/dom/issues/544 apparently this is a thing. But hey, why tell us, it's not like anyone cares.

**Comment by e-cloud** [comment]
@mhevery any progress on the design? It's has been such a while without further update.

**Comment by lazarljubenovic** [comment]
My guess is that the answer is: "it will all be much simpler once Ivy is stable".

**Comment by chaosmonster** [comment]
Now that Ivy is almost around the corner. Is there any thoughts on this issue?

**Comment by mhevery** [comment]
We will not work on this until ivy is fully landed and default render. Preview is in v8 and default is in v9 https://blog.angular.io/a-plan-for-version-8-0-and-ivy-b3318dfc19f7 so not before than.

**Comment by stupidawesome** [comment]
Those interested cold event streams should check out this small library I'm working called [NgObservable](https://github.com/zodiac-team/zodiac-ui/tree/master/libs/ng-observable) that I believe fits the bill.

It includes:
- Aot compatible observable extensions for `HostListener`, `(event)` bindings and lifecycle hooks.
- Easy subscription management with automatic cleanup.
- Advanced state management with total control over change detection (zones, OnPush are made redundant)
- [And much more](https://dev.to/michaelmuscat/create-observable-angular-components-with-ngobservable-2424)

Cold event streams are achieved using a special subject called `InvokeSubject`. This takes the place of your usual method handler and transforms it into an observable stream. Combined with "noop" mode for NgZone, event bindings no longer trigger change detection and you can manually configure every component's change detection cycle by manually calling `state.next()` when using this library's State management solution. Combined with observables it looks something like this:

```ts
@Component({
    selector: "app-hello",
    template: `
        <button (click)="btnClick.next($event)">Click me</button>
        <p>Coords:</p>
        <pre>{{ '{' }} x: {{coords?.screenX || 0}}, y: {{coords?.screenY || 0}} {{ '}' }}</pre>
    `,
    styleUrls: ["./hello.component.scss"],
    viewProviders: [StateFactory, useStateChangeStrategy(StateChangeStrategy.DETACH)]
})
export class HelloComponent extends NgObservable<HelloProps> {
    @Output()
    readonly btnClick = new InvokeSubject<Event>()
    
    @HostListener("mousemove", ["$event"])
    public mousemove = new InvokeSubject<MouseEvent>()

    readonly coords: MouseEvent

    constructor(
        @Self() stateFactory: StateFactory<HelloComponent>
    ) {
        super()
        const state = stateFactory.create(this)
        this.sink = stream(state)(setCoords(this.mousemove))
    }
}
```

It's still in the works but I'm keen to get feedback. You can [grab it on NPM](https://www.npmjs.com/package/@zodiac-ui/ng-observable).

**Comment by d3lm** [comment]
Hey everyone 👋 I wanted to throw in a small library called `ngx-template-streams`. It lets you define **event streams in templates** using a simple custom DSL that is inspired by the requirements and design by @mhevery.

You can find the library here https://github.com/typebytes/ngx-template-streams.

Essentially in your template you can write something like

```
<button (*click)="clicks$; $event = 'custom payload'">Click Me</button>
```

where overwriting `$event` is optional. The default is just `$event`.

In your component  you can then define an `ObservableEvent` like so:

```
@Component({...})
export class MyComponent implements OnInit {
  @ObservableEvent()
  clicks$: Observable<any>;

  ngOnInit() {
    // we can either manually subscribe or use the async pipe
    this.clicks$.subscribe(console.log);
  }
}
```

That's it.

What the library does is basically hooking into the build process and applying HTML transformations and AST-based TS transformations to transform the custom DSL into something that Angular understands. So the code is simply transformed into a `Subject` and in the template we use that subject to `next` the payload.

Please give it a try and let us know how it goes and if this is any helpful.

**Comment by jinder** [comment]
Any movement on this now that v10 has shipped and Ivy seems stable?

**Comment by mhevery** [comment]
> Any movement on this now that v10 has shipped and Ivy seems stable?

This is not high on our priority list, so it will be a while.

**Comment by hiepxanh** [comment]
I think this is a nice feature, hope it will be in the next release

**Comment by waynon** [comment]
Any update? It's imperative to me.

**Comment by kyro38** [comment]
Take that comment for +1 support for this feature :)

**Comment by VladimirCreator** [comment]
😔
